### PR TITLE
Re-raising exception on deliver (fixes #49)

### LINKED ISSUE #49: Relay deliver eats connection error without reraising the exception
This line:

https://github.com/moggers87/salmon/blob/master/salmon/server.py#L128

shouldn't it reraise the exception?

Otherwise the email is never delivered and no indication of this fact is given to the sender.

### PR DESCRIPTION
**Pull request by DrDub**
In python3 it would be nice to chain it with smtplib.SMTPException.

### CODE REVIEW

**Comment by DrDub** [comment]
Most intriguing. All these new errors on Travis come from this change?
I guess I'll have t expand this PR then.

**Comment by moggers87** [comment]
Yeah, those tests will need fixing. You'll also need to create a test for the new behaviour

**Comment by DrDub** [comment]
OK, now it works and added an extra test case for the behavior (test_relay_asserts_no_relay). 
It launches a smtp.DebuggingServer in a separate process at setup.
Interestingly, before the failing tests were not really "testing" that much because it was just hitting this exception and returning silently. I believe highlighting this (and fixing the tests) to be the most important aspect of this change.

File: tests/salmon_tests/server_tests.py [hunk 1]
```diff
@@ -124,38 +125,45 @@ def test_Relay_asserts_ssl_options():
...
+    """Relay raises socket.error if there is no relay"""
+    relay = server.Relay("localhost", port=1234)
+
+    with assert_raises(socket.error):
+        relay.deliver(test_mail_response_plain_text())
+
 @patch("salmon.server.smtplib.SMTP")
 def test_relay_smtp(client_mock):
-    relay = server.Relay("localhost", port=8899)
+    relay = server.Relay("localhost", port=7899)
```
Comment (moggers87) [inline tests/salmon_tests/server_tests.py]: This is an unnecessary change: we mock out `smtplib.SMTP` already.

File: tests/salmon_tests/testing_tests.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 from .setup_env import setup_salmon_dirs, teardown_salmon_dirs
 
 
-relay = relay(port=8899)
+relay = relay(port=7899)
```
Comment (moggers87) [inline tests/salmon_tests/testing_tests.py]: Why was the port number changed?

File: tests/salmon_tests/server_tests.py [hunk 3]
```diff
@@ -164,7 +172,7 @@ def test_relay_smtp_ssl(client_mock):
 def test_relay_deliver_mx_hosts(query):
     query.return_value = [Mock()]
     query.return_value[0].exchange = "localhost"
-    relay = server.Relay(None, port=8899)
+    relay = server.Relay(None, port=7899)
```
Comment (moggers87) [inline tests/salmon_tests/server_tests.py]: We should mock out smtplib here too

File: tests/salmon_tests/server_tests.py [hunk 4]
```diff
@@ -191,7 +199,7 @@ def test_relay_resolve_relay_host(query):
 
 
 def test_relay_reply():
-    relay = server.Relay("localhost", port=8899)
+    relay = server.Relay("localhost", port=7899)
```
Comment (moggers87) [inline tests/salmon_tests/server_tests.py]: Ditto

File: tests/salmon_tests/__init__.py [hunk 5]
```diff
@@ -1,17 +1,24 @@
...
 
 dirs = ["run", "logs"]
+DUMMY_RELAY=None
 
 
 def setup_package():
     for path in dirs:
         os.mkdir(path)
+    DUMMY_RELAY=subprocess.Popen("python -m smtpd -n -c DebuggingServer localhost:7899 > logs/smtpd.log", shell=True)
+    time.sleep(10)
```
Comment (moggers87) [inline tests/salmon_tests/__init__.py]: I don't think we need this - see my comments on the relay tests

File: tests/salmon_tests/testing_tests.py (hunk 2 above)
Comment (DrDub) [inline tests/salmon_tests/testing_tests.py]: Sorry, that port was in use in my machine, 8899 is not such an unusual port.

File: tests/salmon_tests/__init__.py (hunk 5 above)
Comment (DrDub) [inline tests/salmon_tests/__init__.py]: It all boils down to this comment on the original code:

https://github.com/moggers87/salmon/blob/master/tests/salmon_tests/server_tests.py#L126
"this test actually delivers to a test server"

My understanding of the lamson testing code is that it assumes you're running a stmp at 8899. By virtue of silently failing if there is none, all these tests seem to be correct, when they were not.

The relay was not being tested. It was failing and continuing silently.

My initial fix was to set a "in_test" argument in Relay to mimic that behavior, but that didn't work with command_tests.

How do you want to reuse the smtplib mock?

Comment (moggers87) [inline tests/salmon_tests/__init__.py]: That's an old comment, it should be removed ☺

As far as I can see, all the tests that fail without a test server are all testing **sending**, not receiving. 

You can use mock like tihs:
```
@patch("salmon.server.smtplib.SMTP")
def test_server_test(client_mock):
    # test code
```

This is the documentation for the mock library we're using: http://www.voidspace.org.uk/python/mock/

**Comment by DrDub** [comment]
OK, I think this could be it. Thanks for your help. 

If this is a go, let me know and I'll rebase it.

**Comment by moggers87** [comment]
Looks good to me, if you can undo your changes to the port numbers[1] used in tests and remove the unused imports in `tests/salmon_tests/__init__.py` then this is good to be merged.

Sorry for the delay in reviewing your latest changes!

[1]: I'll open a separate bug for dealing with port number clashes during testing. No matter what we choose, it will be in use on someone's system somewhere so let's do things properly :)

**Comment by moggers87** [comment]
Closing this PR as I've fixed this issue in master by not catching the exception at all. There's no real reason to catch an exception, log it, and then re-raise it.
